cmd/up.go

```diff
@@ -61,6 +67,73 @@ func UpCommand() cli.Command {
 	}
 }
 
+func doUpgradeLegacyCluster(ctx context.Context, kubeCluster *cluster.Cluster, fullState *cluster.RKEFullState, k8sWrapTransport k8s.WrapTransport) error {
+	if _, err := os.Stat(kubeCluster.LocalKubeConfigPath); os.IsNotExist(err) {
+		// there is no kubeconfig. This is a new cluster
+		logrus.Debug("[state] local kubeconfig not found, this is a new cluster")
+		return nil
+	}
+	if _, err := os.Stat(kubeCluster.StateFilePath); err == nil {
+		// this cluster has a previous state, I don't need to upgrade!
+		logrus.Debug("[state] previous state found, this is not a legacy cluster")
+		return nil
+	}
+	// We have a kubeconfig and no current state. This is a legacy cluster or a new cluster with old kubeconfig
+	// let's try to upgrade
+	log.Infof(ctx, "[state] Possible legacy cluster detected, trying to upgrade")
+	if err := cluster.RebuildKubeconfig(ctx, kubeCluster); err != nil {
+		return err
+	}
+	recoveredCluster, err := cluster.GetStateFromKubernetes(ctx, kubeCluster.LocalKubeConfigPath, kubeCluster.K8sWrapTransport)
+	if err != nil {
+		return err
+	}
+	// if we found a recovered cluster, we will need override the current state
+	if recoveredCluster != nil {
+		recoveredCerts, err := cluster.GetClusterCertsFromKubernetes(ctx, kubeCluster.LocalKubeConfigPath, kubeCluster.K8sWrapTransport, kubeCluster.EtcdHosts)
+		if err != nil {
+			return err
+		}
+		fullState.CurrentState.RancherKubernetesEngineConfig = recoveredCluster.RancherKubernetesEngineConfig.DeepCopy()
+		fullState.CurrentState.CertificatesBundle = recoveredCerts
+
+		// we don't want to regenerate certificates
+		fullState.DesiredState.CertificatesBundle = recoveredCerts
+		if err = fullState.WriteStateFile(ctx, kubeCluster.StateFilePath); err != nil {
+			return err
+		}
+		return cluster.RemoveLegacyStateFromKubernets(ctx, kubeCluster.LocalKubeConfigPath, kubeCluster.K8sWrapTransport)
+	}
+
+	return nil
+}
+
+func ClusterInit(ctx context.Context, rkeConfig *v3.RancherKubernetesEngineConfig, configDir string, k8sWrapTransport k8s.WrapTransport) error {
+	log.Infof(ctx, "Initiating Kubernetes cluster")
+	stateFilePath := cluster.GetStateFilePath(clusterFilePath, configDir)
+	rkeFullState, _ := cluster.ReadStateFile(ctx, stateFilePath)
+	kubeCluster, err := cluster.InitClusterObject(ctx, rkeConfig, clusterFilePath, configDir)
+	if err != nil {
+		return err
+	}
+
+	err = doUpgradeLegacyCluster(ctx, kubeCluster, rkeFullState, k8sWrapTransport)
+	if err != nil {
+		log.Warnf(ctx, "[state] can't fetch legacy cluster state from Kubernetes")
+	}
+
+	fullState, err := cluster.RebuildState(ctx, &kubeCluster.RancherKubernetesEngineConfig, rkeFullState, clusterFilePath, configDir)
+	if err != nil {
+		return err
+	}
+
+	rkeState := cluster.RKEFullState{
+		DesiredState: fullState.DesiredState,
+		CurrentState: fullState.CurrentState,
+	}
+	return rkeState.WriteStateFile(ctx, stateFilePath)
+}
+
 func ClusterUp(
 	ctx context.Context,
 	rkeConfig *v3.RancherKubernetesEngineConfig,
```

Comment:
I guess we don't need rkeConfig as it's supposed to be initialized with init?

Reply:
fixed